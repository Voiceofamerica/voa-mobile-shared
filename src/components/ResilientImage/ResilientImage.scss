
@import '../../commonStyles/index.scss';

.resilientImage {
  position: relative;
  background: black;

  .imageContent {
    object-fit: cover;
    width: 100%;
    height: 100%;

    &.containImage {
      object-fit: contain;
      height: auto;
      max-height: 100%;
    }
  }

  .childrenArea {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .spinner {
    color: $voa-blue;
    height: 80%;
    max-height: 100px;
    margin-top: -1%;
  }
}
