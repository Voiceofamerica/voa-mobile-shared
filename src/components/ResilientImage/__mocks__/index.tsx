
import * as React from 'react'

export default class ResilientImage extends React.Component {
  render () {
    const { children, ...props } = this.props
    return (
      <div {...props}>{children}</div>
    )
  }
}
