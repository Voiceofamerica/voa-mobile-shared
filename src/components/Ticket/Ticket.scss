
@import '../../commonStyles/index.scss';

$image-width: 30vw;

@keyframes imageSlide {
  from {
    width: 0;
  }
  to {
    width: $image-width;
  }
}

.ticket {
  @include touchable();

  flex: 1;
  display: flex;
  flex-direction: row;
  background: $light-grey;
  box-shadow: 0 1vw 2vw rgba(0, 0, 0, 0.8);
  margin: 1.5vw;
  border-radius: 1vw;
  height: 20vw;
  font-size: 3vw;
  overflow: hidden;
  user-select: none;

  .imageArea {
    width: 0;
    position: relative;
    visibility: hidden;
    background-size: cover;
    background-position: center;

    &.show {
      width: $image-width;
      visibility: visible;
      animation-duration: 0.3s;
      animation-name: imageSlide;
      animation-iteration-count: 1;
    }

    .iconOverlay {
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }

  .textContent {
    flex: 2;
    display: flex;
    flex-direction: column;
    padding: 1vw 1.5vw;
    
    .titleText {
      flex: 1;
      overflow: hidden;
      position: relative;
      font-weight: 'bold';
      font-size: 1.1em;
      top: 0.5vw;
      
      .fadeOut {
        position: absolute;
        bottom: 0;
        left: 0;
        right: 0;
        height: 5vw;
        background: linear-gradient(transparent, $light-grey 90%, $light-grey 100%);
      }
    }

    .minorText {
      color: #444;
      text-align: right;
      font-size: 0.8em;
      border-top: 2px solid $light-grey;
      background: $light-grey;
    }
  }
}
