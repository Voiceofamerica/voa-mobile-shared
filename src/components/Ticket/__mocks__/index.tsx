
import * as React from 'react'

export default (props) => (
  React.createElement('ticket', { props }, props.children)
)

export const TICKET_HEIGHT = 100
