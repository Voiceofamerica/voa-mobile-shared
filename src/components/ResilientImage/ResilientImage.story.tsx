
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import ResilientImage from './ResilientImage'

const imageUrl = 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w650_r1.png'

storiesOf('ResilientImage', module)
  .add('display', () => (
    <ResilientImage src={imageUrl} />
  ))
  .add('display without image', () => (
    <ResilientImage src='asdf' />
  ))
