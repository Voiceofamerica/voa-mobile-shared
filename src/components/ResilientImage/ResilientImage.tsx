
import * as React from 'react'

import { waitUntilOnline } from '../../helpers/resilience'

import Spinner from '../Spinner'

import { resilientImage, imageContent, childrenArea, containImage, spinner } from './ResilientImage.scss'

export interface Props {
  src: string | null
  showSpinner?: boolean
  className?: string
  alwaysShow?: boolean
  style?: React.CSSProperties
  defaultSrc?: string
  onLoadDone?: () => void
  onClick?: () => void
  contain?: boolean
}

export type ImageStatus = 'loading' | 'loaded' | 'error'

export interface State {
  shouldRender?: boolean
  imgUrl?: string
  imageStaus: ImageStatus
}

const noop = () => null
const IMAGE_FETCH_RETRY_RATE = 2000

class ReilientImage extends React.Component<Props, State> {
  state: State = {
    imageStaus: 'loading',
  }

  mounted: boolean = false

  readonly abortController = new AbortController()

  componentWillMount () {
    this.mounted = true
    const shouldRender = navigator.onLine || this.props.alwaysShow
    this.setState({
      shouldRender,
    })

    if (shouldRender) {
      this.tryFetchImage()
    }
  }

  componentWillUnmount () {
    this.mounted = false
    this.abortController.abort()
  }

  componentWillReceiveProps (nextProps: Props) {
    if (nextProps.src !== this.props.src) {
      this.setImageStatus('loading')

      const shouldRender = navigator.onLine || this.props.alwaysShow
      this.setState({
        shouldRender,
      })

      if (shouldRender) {
        this.tryFetchImage()
          .catch(err => {
            console.warn(`Something went wrong loading ${nextProps.src}`, err)
          })
      }
    }
  }

  tryFetchImage = async () => {
    const { src } = this.props

    if (!src || !this.mounted) {
      return
    }

    if (!navigator.onLine) {
      waitUntilOnline().then(() => this.mounted && this.tryFetchImage())
      return
    }

    const { onLoadDone = noop } = this.props

    try {
      const blobUrl = await fetch(src, {
          method: 'GET',
          signal: this.abortController.signal,
        })
        .then(res => res.blob())
        .then(blob => URL.createObjectURL(blob))

      this.setImageStatus('loaded', blobUrl)
      onLoadDone()

    } catch (err) {
      this.setImageStatus('error')
      setTimeout(this.tryFetchImage, IMAGE_FETCH_RETRY_RATE)
    }
  }

  setImageStatus = (imageStaus: ImageStatus, imgUrl?: string) => {
    if (!this.mounted) {
      return
    }

    this.setState({
      imageStaus,
      imgUrl,
    })
  }

  renderOffline = () => (
    <div style={{ width: 0, height: 0 }} />
  )

  renderContent = () => {
    const { imageStaus, imgUrl } = this.state
    const { defaultSrc = require('./imagedefault.gif'), contain } = this.props

    const isLoaded = imageStaus === 'loaded'
    const srcToUse = isLoaded && imgUrl
                   ? imgUrl
                   : defaultSrc

    const className = contain ? `${imageContent} ${containImage}` : imageContent

    return (<img src={srcToUse} className={className} />)
  }

  renderSpinner = () => {
    const { showSpinner } = this.props
    const { imageStaus } = this.state

    const isLoaded = imageStaus === 'loaded'

    if (isLoaded || !showSpinner) {
      return null
    }

    return (
      <div className={childrenArea}>
        <Spinner className={spinner} />
      </div>
    )
  }

  renderChildren = () => {
    const { children } = this.props

    return (
      <div className={childrenArea}>
        {children}
      </div>
    )
  }

  render () {
    const { shouldRender = navigator.onLine || this.props.alwaysShow } = this.state
    if (!shouldRender) {
      return this.renderOffline()
    } else {
      const {
        className = '',
        style,
        onClick = () => null,
      } = this.props

      return (
        <div className={`${resilientImage} ${className}`} style={style} onClick={ev => onClick()}>
          { this.renderContent() }
          { this.renderSpinner() }
          { this.props.children ? this.renderChildren() : null }
        </div>
      )
    }
  }
}

export default ReilientImage
