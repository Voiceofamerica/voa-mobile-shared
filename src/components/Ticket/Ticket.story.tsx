
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import ArticleBlurb from '../../types/ArticleBlurb'

import Ticket from './Ticket'

const testArticle = {
  id: 1,
  imageUrl: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w650_r1.png',
  pubDate: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const testArticle2 = {
  imageUrl: 'https://gdb.voanews.com/01D5530E-3642-4CBF-957E-AE990ABE9167_cx0_cy14_cw0_w650_r1.jpg',
  pubDate: '21:55',
  title: 'At least seven people were killed in a police vehicle in Balochistan, Pakistan',
}

storiesOf('Ticket', module)
  .add('display', () => (
    <div style={{ display: 'flex', flexDirection: 'row' }}>
      <Ticket onPress={action('Card.onPress')} { ...testArticle } />
    </div>
  ))
  .add('display with icon', () => (
    <div style={{ display: 'flex', flexDirection: 'row' }}>
      <Ticket
        onPress={action('Card.onPress')}
        { ...testArticle }
        icon={<img src={require('./headphones.svg')} />}
      />
    </div>
  ))
