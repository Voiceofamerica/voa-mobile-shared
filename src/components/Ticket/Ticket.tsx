
import * as React from 'react'

import ResilientImage from '../ResilientImage'

import { ticket, imageArea, show, content, textContent, minorText, titleText, fadeOut } from './Ticket.scss'

export interface Props {
  title: JSX.Element|string
  onPress?: () => void
  description?: string
  imageUrl?: string
  minorText?: JSX.Element|string
  icon?: JSX.Element
}

export interface State {
  showImage: boolean
}

class Ticket extends React.Component<Props, State> {
  state = {
    showImage: false,
  }

  onImageLoaded = () =>
    this.setState({
      showImage: true,
    })

  render () {
    const { onPress, imageUrl, title, description, minorText: minor, icon } = this.props
    const { showImage } = this.state

    const imageClass = showImage
                     ? `${imageArea} ${show}`
                     : imageArea

    return (
      <div className={ticket} onClick={onPress}>
        <ResilientImage src={imageUrl} alwaysShow className={imageClass} onLoadDone={this.onImageLoaded}>
          {icon}
        </ResilientImage>
        <div className={content}>
          <div className={textContent}>
            <div className={titleText}>
              {title}
            </div>
            {description}
            <div className={fadeOut}/>
          </div>
          <div className={minorText}>
            {minor}
          </div>
        </div>
      </div>
    )
  }
}

export default Ticket
