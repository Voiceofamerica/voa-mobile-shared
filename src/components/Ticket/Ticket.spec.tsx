
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import Ticket, { Props, State } from './Ticket'
import ResilientImage from '../ResilientImage'

jest.mock('../ResilientImage')
jest.mock('../SvgIcon')

describe(`<${Ticket.name} />`, () => {
  describe('shapshots', () => {
    it('should render with the given title', () => {
      const element = create((
        <Ticket title='myTitle' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given image', () => {
      const element = create((
        <Ticket title='myTitle' imageUrl='myImage' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given minor text', () => {
      const element = create((
        <Ticket title='myTitle' minorText='This is some minor text' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given icon shown', () => {
      const element = create((
        <Ticket title='myTitle' icon={<i className='mdi mdi-monitor' />} />
      ))

      expect(element).toMatchSnapshot()
    })
  })

  describe('onPress', () => {
    it('should be called whenever the element is clicked', () => {
      const onPress = jest.fn()

      const item = shallow((
        <Ticket title='myTitle' onPress={onPress} />
      ))

      item.simulate('click')

      expect(onPress).toHaveBeenCalledTimes(1)
    })

    it('should not throw an error if not provided and the element is clicked', () => {
      const item = shallow((
        <Ticket title='myTitle' />
      ))

      item.simulate('click')
    })
  })

  describe('image onLoadDone', () => {
    it('should not show the image if it is not yet loaded', () => {
      const item = shallow<Props, State>((
        <Ticket title='myTitle' />
      ))

      expect(item.state().showImage).toBe(false)
    })

    it('should show the image if it is done loading', () => {
      const item = shallow<Props, State>((
        <Ticket title='myTitle' />
      ))

      item.find(ResilientImage).props().onLoadDone()

      expect(item.state().showImage).toBe(true)
    })
  })
})
