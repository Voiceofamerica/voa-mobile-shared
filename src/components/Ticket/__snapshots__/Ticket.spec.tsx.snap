// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> shapshots should render with the given icon shown 1`] = `
<div
  className="ticket"
  onClick={undefined}
>
  <div
    alwaysShow={true}
    className="imageArea"
    onLoadDone={[Function]}
    src={undefined}
  >
    <i
      className="mdi mdi-monitor"
    />
  </div>
  <div
    className="content"
  >
    <div
      className="textContent"
    >
      <div
        className="titleText"
      >
        myTitle
      </div>
      <div
        className="fadeOut"
      />
    </div>
    <div
      className="minorText"
    />
  </div>
</div>
`;

exports[`<Card /> shapshots should render with the given image 1`] = `
<div
  className="ticket"
  onClick={undefined}
>
  <div
    alwaysShow={true}
    className="imageArea"
    onLoadDone={[Function]}
    src="myImage"
  />
  <div
    className="content"
  >
    <div
      className="textContent"
    >
      <div
        className="titleText"
      >
        myTitle
      </div>
      <div
        className="fadeOut"
      />
    </div>
    <div
      className="minorText"
    />
  </div>
</div>
`;

exports[`<Card /> shapshots should render with the given minor text 1`] = `
<div
  className="ticket"
  onClick={undefined}
>
  <div
    alwaysShow={true}
    className="imageArea"
    onLoadDone={[Function]}
    src={undefined}
  />
  <div
    className="content"
  >
    <div
      className="textContent"
    >
      <div
        className="titleText"
      >
        myTitle
      </div>
      <div
        className="fadeOut"
      />
    </div>
    <div
      className="minorText"
    >
      This is some minor text
    </div>
  </div>
</div>
`;

exports[`<Card /> shapshots should render with the given title 1`] = `
<div
  className="ticket"
  onClick={undefined}
>
  <div
    alwaysShow={true}
    className="imageArea"
    onLoadDone={[Function]}
    src={undefined}
  />
  <div
    className="content"
  >
    <div
      className="textContent"
    >
      <div
        className="titleText"
      >
        myTitle
      </div>
      <div
        className="fadeOut"
      />
    </div>
    <div
      className="minorText"
    />
  </div>
</div>
`;
